### PR TITLE
feature: 2 algorithms Website-specific workers and Round robin +DNS

### PR DESCRIPTION
**Pull request by deputinizer**
They use a lot less resources, since eg.
1 MB buffer is assigned to the same URL
and other 1 kB buffer is assigned to another, constant URL

Also spamming goroutines with httpClient.DoContext was unwise.
Runtime goroutine switching took a lot of CPU.

Currently, i see a lot less RAM+CPU usage after removing task-issuing channel.

### CODE REVIEW

**Comment by deputinizer** [comment]
Also, added stats `localhost:8049/downloaded`
```csv
 4244.084 MB	https://www.moex.com
 2412.250 MB	https://mail.ru
 2330.066 MB	https://www.wildberries.ru
 2307.288 MB	https://abcobmen.com
 2106.000 MB	https://www.rambler.ru
 1780.000 MB	https://www.yaplakal.com
  595.322 MB	https://www.avito.ru
  542.826 MB	https://any.yandex.ru
  529.330 MB	https://montreal.mid.ru
```

**Comment by deputinizer** [comment]
<details>
  <summary>Main CPU abuser now [spoiler]</summary>
  
image
  
</details>

**Comment by deputinizer** [comment]
<details>
  <summary>We should nuke this buffer with Bayraktars</summary>

image
  
</details>

**Comment by erkexzcx** [comment]
I don't like the idea to bring per-website workers. There was an issue that even 1 worker (per website) used all traffic, so workers pool was the way to fix it.

I understand that this way we could save more resources and make this app more performant, but what can we offer for users where even 1 worker per website is too much?

Shall we ignore such users and go ahead with workers-per-website approach?

Or shall we still use the same "workers pool" idea and divide workers equally to each website (rouding workers number so each website has the same amount of workers)?

Opinions? :)

**Comment by deputinizer** [comment]
> Shall we ignore such users and go ahead with workers-per-website approach?

Yes

bytes/s > packets/s

Or maybe add 2 options - your worker issuing channel algorithm and worker-per-website algorithm.

**Comment by deputinizer** [comment]
Now fasthttp is able to `io.Discard` :smiley: https://github.com/deputinizer/fasthttp/commit/80e5eb0ab1f4788e5b0e375dd3f4d03d6879ecab

**Comment by deputinizer** [comment]
<details>
  <summary>Discards like a boss (heap pprof)</summary>
  
image

  
</details>

**Comment by deputinizer** [comment]
<details>
  <summary>There's something wrong i think (CPU pprof)</summary>
  
image

</details>

**Comment by deputinizer** [comment]
So, looks like:
- round robin prioritizes slower websites and spam with `net.Dial`
- parallel goroutines (few per website) prioritize wasting throughput

You can choose which task issuing algorithm to use with a flag:
`-roundrobin=true`

**Comment by deputinizer** [comment]
Ok, so stats now work with `-roundrobin=true`
```
   43.239 MB	https://abcobmen.com
   33.885 MB	https://www.rambler.ru
   22.483 MB	https://mail.ru
   19.784 MB	https://pikabu.ru
   16.967 MB	https://www.moex.com
    6.215 MB	https://montreal.mid.ru
    2.487 MB	https://www.avito.ru
    1.695 MB	https://www.yaplakal.com
```

For comparison `-roundrobin=false`
```
  207.586 MB	https://www.moex.com
  128.076 MB	https://www.yaplakal.com
  119.014 MB	https://www.wildberries.ru
  118.598 MB	https://mail.ru
  113.556 MB	https://www.rambler.ru
  109.937 MB	https://abcobmen.com
   50.105 MB	https://www.avito.ru
   32.551 MB	https://any.yandex.ru
   26.804 MB	https://pikabu.ru
   25.712 MB	https://montreal.mid.ru
   19.033 MB	https://ww-pay.net
   14.681 MB	https://superchange.net
```

**Comment by deputinizer** [comment]
I suggest `-roundrobin=false` by default
since new StopPropaganda's users are going to be astonished after seeing 10 Gbit/s on their machines on first try. :satisfied:

**Comment by deputinizer** [comment]
Ok, this DNS client replaces `/etc/resolv.conf` in a proper way (no `net.Resolver.Dial` hacks), just a lot of copy paste.
I had to `ctrl+C` `ctrl+V` a lot of unix DNS client, because 90% of the code was private :cry:

File: internal/stoppropaganda/customresolver/customresolver.go [hunk 1]
```diff
@@ -6,11 +6,34 @@ import (
...
 
 var DnsCache *cache.Cache
 
+func MakeDNSConfig() (conf *spdnsclient.SPDNSConfig) {
+	conf = &spdnsclient.SPDNSConfig{
+		Ndots:    1,
+		Timeout:  5 * time.Second,
+		Attempts: 2,
+	}
+	conf.Servers = targets.ReferenceDNSServersForHTTP
```
Comment (deputinizer) [inline internal/stoppropaganda/customresolver/customresolver.go]: Here is the actual fake `/etc/resolv.conf`

**Comment by erkexzcx** [comment]
Regarding https://github.com/erkexzcx/stoppropaganda/pull/103/commits/30f0113715b9b1cc9b8f9ada345cd1588a64ac2a - this was not needed. All the blockings were done by `if !website.allowedToRun()` part.

**Comment by erkexzcx** [comment]
Works great on Windows 10 :+1:

**Pull request merged**